UCBChangesToScriptPics.js

```diff
@@ -0,0 +1,486 @@
+
+BlockMorph.prototype.userMenu = function () {
+    var menu = new MenuMorph(this),
+        world = this.world(),
+        myself = this,
+        hasLine = false,
+        shiftClicked = world.currentKey === 16,
+        proc = this.activeProcess(),
+        top = this.topBlock(),
+        vNames = proc && proc.context && proc.context.outerContext ?
+                proc.context.outerContext.variables.names() : [],
+        alternatives,
+        field,
+        rcvr;
+
+    function addOption(label, toggle, test, onHint, offHint) {
+        menu.addItem(
+            [
+                test ? new SymbolMorph(
+                    'checkedBox',
+                    MorphicPreferences.menuFontSize * 0.75
+                ) : new SymbolMorph(
+                    'rectangle',
+                    MorphicPreferences.menuFontSize * 0.75
+                ),
+                localize(label)
+            ],
+            toggle,
+            test ? onHint : offHint
+        );
+    }
+
+    function renameVar() {
+        var blck = myself.fullCopy();
+        blck.addShadow();
+        new DialogBoxMorph(
+            myself,
+            myself.userSetSpec,
+            myself
+        ).prompt(
+            "Variable name",
+            myself.blockSpec,
+            world,
+            blck.doWithAlpha(1, () => blck.fullImage()), // pic
+            InputSlotMorph.prototype.getVarNamesDict.call(myself)
+        );
+    }
+
+    menu.addItem(
+        "help...",
+        'showHelp'
+    );
+
+    /* UCB Script Pic Edit:
+     *  - Incorporated new method to userMenu to integrate scriptPic's and XML code.
+     *  - Code credited to Snap! Forum user Dardoro.
+     */
+    if (this.isTemplate || !this.definition) {
+		menu.addLine();
+		menu.addItem("script pic and xml...", 
+			() => {
+				const inbDelim = "Snap\tBlocks\tEmbedded"; //in-band XML delimeter
+				function crc32(str, crc) {
+					let table = [...Array(256).keys()].map(it => 
+						[...Array(8)].reduce((cc) => 
+						(cc & 1) ? (0xedb88320 ^ (cc >>> 1)) : (cc >>> 1), it)
+						);
+						crc = [...str].reduce((crc, ch) => { 
+							return (crc >>> 8) ^ table[(crc ^ ch.charCodeAt(0)) & 0xff]
+						}, (crc ||= 0) ^ (-1));
+						return ( crc ^ (-1) ) >>> 0;
+					};
+				function arr2Str(arr) { 
+					return arr.reduce((res, byte) => res + String.fromCharCode(byte), '');
+				};
+				function int2BStr(val) { 
+					return arr2Str(Array.from(new Uint8Array(new Uint32Array( [val] ).buffer)).reverse());
+				};
+				function buildChunk(type , data) { 
+					let res = type + data; 
+					return int2BStr(data.length) + res + int2BStr(crc32(res));
+				};
+				function SaveData(dataURL, name) { 
+					return Object.assign(document.createElement("a"), 
+					{download: name, href: dataURL}).click(); 
+				};
+					
+				var ide = this.parentThatIsA(IDE_Morph),
+					top = this.definition || this.topBlock();
+				let xml = unescape(encodeURIComponent(ide.serializer.serialize(top))),
```

Comment:
what happens to dependencies referenced in the script, i.e. custom blocks? This is a loose end on my side, so I'm afraid that it'll cause red "undefined" blocks when folks file it in.

Reply:
Yes, I hadn't thought about this, but referencing a custom block within a script does bring back an "undefined" block. To your point below, we will most likely have to scan for the existence of a custom block before serializing. I'll look into this more in depth.